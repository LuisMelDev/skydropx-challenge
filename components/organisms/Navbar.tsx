import React from "react";
import Image from "next/image";
import Link from "next/link";

export const Navbar = () => {
  return (
    <header>
      <nav className="w-full h-14 bg-white flex items-center px-4 shadow-sm">
        <Link passHref href="/">
          <a>
            <div className="relative w-32 h-12">
              <Image
                src="https://skydropx.com/assets/landing5/skydropx_logo-4fb27c0601c3bccddd15cad09e612eb0d777dcdbcebae56c0382a8fe2978dfa9.svg"
                alt="Logo"
                layout="fill"
              />
            </div>
          </a>
        </Link>
      </nav>
    </header>
  );
};
