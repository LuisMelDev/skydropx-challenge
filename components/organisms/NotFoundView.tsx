import Link from "next/link";
import React from "react";
import { Button } from "..";

export const NotFoundView = () => {
    return (
        <div className='h-full w-full flex flex-col justify-center items-center mt-10'>
            <h1 className='text-9xl font-extrabold text-gray-600 tracking-widest'>
                404
            </h1>
            <div className='bg-blue-600 text-white px-2 text-sm rounded rotate-12 absolute'>
                Page Not Found
            </div>
            <div className='mt-5'>
                <span className='relative block '>
                    <Button className='px-5 py-2 rounded-md text-white bg-blue-500 shadow-md'>
                        <Link href='/' passHref>
                            <a>Ir al Inicio</a>
                        </Link>
                    </Button>
                </span>
            </div>
        </div>
    );
};
