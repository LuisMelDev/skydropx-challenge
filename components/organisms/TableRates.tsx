import { MouseEventHandler } from "react";
import { ItemTable, TableHead } from "..";

const heads = [
    "Paqueteria",
    "Servicio",
    "Entrega Estimada",
    "Precio",
    "Recomendacion",
];

interface TableRatesProps {
    rates: any;
    onClickRate: (id:string) => void
}

export const TableRates = ({ rates,onClickRate }: TableRatesProps) => {
    return (
        <div className='-mx-4 sm:-mx-8 px-4 sm:px-8 pt-4 overflow-x-auto'>
            <div className='inline-block min-w-full overflow-hidden'>
                <table className='min-w-full leading-normal'>
                    <TableHead heads={heads} />
                    <tbody>
                        {rates.map((rate: any, index: number) => (
                            <ItemTable
                                key={rate.id}
                                rate={rate}
                                recomended={index === 0}
                                onClick={() => onClickRate(rate.id)}
                            />
                        ))}
                    </tbody>
                </table>
            </div>
        </div>
    );
};
