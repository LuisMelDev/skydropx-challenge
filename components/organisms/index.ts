export * from "./Navbar"
export * from "./ShipmentForm"
export * from "./TableRates"
export * from "./NotFoundView"