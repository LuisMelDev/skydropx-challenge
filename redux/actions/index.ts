export * from './uiAction'
export * from './shipmentActions'
export * from './labelActions'
export * from './ratesActions'