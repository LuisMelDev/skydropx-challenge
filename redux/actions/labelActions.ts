import types from "../types"
import { createLabel } from "services";
import { startLoading, finishLoading, setError, removeError } from './uiAction';

export const startAddLabel = (id: string) => (dispatch:any) => {

    dispatch(startLoading())
    dispatch(removeError())

    createLabel(id)
            .then(resp => {
                dispatch(finishLoading());
                if(resp.data.data.attributes.status === "ERROR"){
                    dispatch(setError(resp.data.data.attributes.error_message[0].message));
                    return;
                }
                dispatch(finishLoading());
                dispatch(addLabel(resp.data.data));
            })
            .catch((error) => {
                dispatch(finishLoading());
                dispatch(setError(error.response.data.message));
                console.log(error);
            });
}


export const addLabel = (payload:any) => ({
    type: types.ADD_LABEL,
    payload: payload
})

export const removeLabel = () => ({
    type: types.REMOVE_LABEL,
})