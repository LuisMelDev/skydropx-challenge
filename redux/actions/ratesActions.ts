import types from "../types"

export const setRates = (payload:any) => ({
    type: types.SET_RATES,
    payload: payload
})

export const removeRates = () => ({
    type: types.REMOVE_RATES,
})