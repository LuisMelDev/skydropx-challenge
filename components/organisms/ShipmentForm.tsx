import React, { useState, ChangeEventHandler, FormEventHandler } from "react";
import { useDispatch } from "react-redux";
import { Button, FormControl } from "..";
import { getBodyShipment } from "utils";
import { startAddShipment } from "redux/actions";

var only_numbers = /^\d+$/;

enum FormErrors {
    ALL_REQUIRED,
    VALIDATION_ERROR,
}

interface ValueShipmentForm {
    zip_from: string;
    zip_to: string;
    weight: string;
    height: string;
    width: string;
    length: string;
}

interface StateError {
    error: FormErrors | null;
}

export const ShipmentForm = () => {
    const [formulario, setFormulario] = useState<ValueShipmentForm>({
        zip_from: "",
        height: "",
        length: "",
        weight: "",
        width: "",
        zip_to: "",
    });
    const dispatch = useDispatch();

    const [error, setError] = useState<StateError>({ error: null });

    const { zip_from, height, length, weight, width, zip_to } = formulario;

    const onChangeForm: ChangeEventHandler<HTMLInputElement> = (e) => {
        e.preventDefault();

        setFormulario({
            ...formulario,
            [e.target.name]: e.target.value,
        });
    };

    const onSubmit: FormEventHandler<HTMLFormElement> = (e) => {
        e.preventDefault();
        setError({ error: null });

        if (
            zip_from.trim() === "" ||
            height.trim() === "" ||
            length.trim() === "" ||
            weight.trim() === "" ||
            width.trim() === "" ||
            zip_to.trim() === ""
        ) {
            setError({ error: FormErrors.ALL_REQUIRED });
            return;
        } else if (
            !only_numbers.test(zip_from.trim()) ||
            !only_numbers.test(height.trim()) ||
            !only_numbers.test(length.trim()) ||
            !only_numbers.test(weight.trim()) ||
            !only_numbers.test(width.trim()) ||
            !only_numbers.test(zip_to.trim())
        ) {
            setError({ error: FormErrors.VALIDATION_ERROR });
            return;
        }

        const body = getBodyShipment(formulario);
        dispatch(startAddShipment(body))
    };

    return (
        <form
            onSubmit={onSubmit}
            className='max-w-lg bg-slate-50 rounded-lg p-5 shadow-lg'
        >
            <h3 className='pb-5 text-4xl font-bold text-center text-gray-700'>
                Cotiza Tu Envio
            </h3>
            <div>
                <FormControl
                    idInput='zip_from'
                    nameInput='zip_from'
                    labelText='Codigo Postal de Origen'
                    onChange={onChangeForm}
                    typeInput='text'
                    value={zip_from}
                />
                <FormControl
                    idInput='zip_to'
                    nameInput='zip_to'
                    labelText='Codigo Postal de Destino'
                    onChange={onChangeForm}
                    typeInput='text'
                    value={zip_to}
                />
                <div className='my-3 text-center'>Medidas del Paquete</div>
                <div className='flex flex-wrap justify-around'>
                    <FormControl
                        idInput='length'
                        nameInput='length'
                        labelText='Largo (cm)'
                        onChange={onChangeForm}
                        typeInput='text'
                        value={length}
                        size='50'
                    />
                    <FormControl
                        idInput='height'
                        nameInput='height'
                        labelText='Alto (cm)'
                        onChange={onChangeForm}
                        typeInput='text'
                        value={height}
                        size='50'
                    />
                    <FormControl
                        idInput='width'
                        nameInput='width'
                        labelText='Ancho (cm)'
                        onChange={onChangeForm}
                        typeInput='text'
                        value={width}
                        size='50'
                    />
                    <FormControl
                        idInput='weight'
                        nameInput='weight'
                        labelText='Peso (kg)'
                        onChange={onChangeForm}
                        typeInput='text'
                        value={weight}
                        size='50'
                    />
                </div>
                <div className='w-full my-2 text-red-500'>
                    {error?.error === FormErrors.ALL_REQUIRED && (
                        <span>Todos los campos son requeridos</span>
                    )}
                    {error?.error === FormErrors.VALIDATION_ERROR && (
                        <span>Todos los campos deben ser numeros</span>
                    )}
                </div>
                <Button
                    type='submit'
                    className='w-full bg-sky-600 rounded-lg p-2 text-white shadow'
                >
                    Enviar Cotizacion
                </Button>
            </div>
        </form>
    );
};
