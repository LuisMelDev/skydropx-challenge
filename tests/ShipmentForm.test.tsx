import React from "react";
import { mount } from "enzyme";
import { ShipmentForm } from "../components/organisms/ShipmentForm";
import { Provider } from "react-redux";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { startAddShipment } from "redux/actions";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock("redux/actions", () => ({
    startAddShipment: jest.fn(),
}));

const initState = {
    auth: {
        uid: "1",
        name: "Fernando",
    },
    ui: {
        loading: false,
        msgError: null,
    },
    notes: {
        active: {
            id: 1234,
            title: "Hola",
            body: "mundo",
            date: 0,
        },
        notes: [],
    },
};

let store = mockStore(initState);

store.dispatch = jest.fn();

describe("Pruebas <ShipmentForm>", () => {
    let wrapper = mount(
        <Provider store={store}>
            <ShipmentForm />
        </Provider>
    );

    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = mount(
            <Provider store={store}>
                <ShipmentForm />
            </Provider>
        );
    });

    test("debe de mostrarse correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    test("Debe mostrar el mensaje 'Todos los campos son requeridos'", () => {
        wrapper.find("form").simulate("submit", { preventDefault() {} });
        expect(wrapper.find("span").text().trim()).toBe(
            "Todos los campos son requeridos"
        );
    });

    test("Debe mostrar el mensaje 'Todos los campos deben ser numeros' y cambiar value de los inputs", () => {
        const value = "asdsa";
        const event = (name: string) => ({ target: { value, name } });

        wrapper
            .find("input[name='zip_from']")
            .simulate("change", event("zip_from"));
        wrapper
            .find("input[name='height']")
            .simulate("change", event("height"));
        wrapper
            .find("input[name='length']")
            .simulate("change", event("length"));
        wrapper
            .find("input[name='weight']")
            .simulate("change", event("weight"));
        wrapper.find("input[name='width']").simulate("change", event("width"));
        wrapper
            .find("input[name='zip_to']")
            .simulate("change", event("zip_to"));
        wrapper.find("form").simulate("submit", { preventDefault() {} });
        expect(wrapper.find("span").text().trim()).toBe(
            "Todos los campos deben ser numeros"
        );
    });
});
