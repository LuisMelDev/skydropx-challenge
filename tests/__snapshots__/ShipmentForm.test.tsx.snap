// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas <ShipmentForm> debe de mostrarse correctamente 1`] = `
<form
  className="max-w-lg bg-slate-50 rounded-lg p-5 shadow-lg"
  onSubmit={[Function]}
>
  <h3
    className="pb-5 text-4xl font-bold text-center text-gray-700"
  >
    Cotiza Tu Envio
  </h3>
  <div>
    <div
      className="full block mb-3"
    >
      <label
        className="text-sm font-light block"
        htmlFor="zip_from"
      >
        Codigo Postal de Origen
      </label>
      <input
        className="bg-slate-200 outline-none focus:bg-slate-50 focus:shadow font-normal h-10 rounded-lg py-1 px-2 text-md block w-full"
        id="zip_from"
        name="zip_from"
        onChange={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      className="full block mb-3"
    >
      <label
        className="text-sm font-light block"
        htmlFor="zip_to"
      >
        Codigo Postal de Destino
      </label>
      <input
        className="bg-slate-200 outline-none focus:bg-slate-50 focus:shadow font-normal h-10 rounded-lg py-1 px-2 text-md block w-full"
        id="zip_to"
        name="zip_to"
        onChange={[Function]}
        placeholder=""
        type="text"
        value=""
      />
    </div>
    <div
      className="my-3 text-center"
    >
      Medidas del Paquete
    </div>
    <div
      className="flex flex-wrap justify-around"
    >
      <div
        className="w-5/12 inline-block mb-3"
      >
        <label
          className="text-sm font-light block"
          htmlFor="length"
        >
          Largo (cm)
        </label>
        <input
          className="bg-slate-200 outline-none focus:bg-slate-50 focus:shadow font-normal h-10 rounded-lg py-1 px-2 text-md block w-full"
          id="length"
          name="length"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
        />
      </div>
      <div
        className="w-5/12 inline-block mb-3"
      >
        <label
          className="text-sm font-light block"
          htmlFor="height"
        >
          Alto (cm)
        </label>
        <input
          className="bg-slate-200 outline-none focus:bg-slate-50 focus:shadow font-normal h-10 rounded-lg py-1 px-2 text-md block w-full"
          id="height"
          name="height"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
        />
      </div>
      <div
        className="w-5/12 inline-block mb-3"
      >
        <label
          className="text-sm font-light block"
          htmlFor="width"
        >
          Ancho (cm)
        </label>
        <input
          className="bg-slate-200 outline-none focus:bg-slate-50 focus:shadow font-normal h-10 rounded-lg py-1 px-2 text-md block w-full"
          id="width"
          name="width"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
        />
      </div>
      <div
        className="w-5/12 inline-block mb-3"
      >
        <label
          className="text-sm font-light block"
          htmlFor="weight"
        >
          Peso (kg)
        </label>
        <input
          className="bg-slate-200 outline-none focus:bg-slate-50 focus:shadow font-normal h-10 rounded-lg py-1 px-2 text-md block w-full"
          id="weight"
          name="weight"
          onChange={[Function]}
          placeholder=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="w-full my-2 text-red-500"
    />
    <button
      className="w-full bg-sky-600 rounded-lg p-2 text-white shadow"
      type="submit"
    >
      Enviar Cotizacion
    </button>
  </div>
</form>
`;
